main.go

```diff
@@ -51,23 +62,73 @@ func main() {
 		cancel()
 	}()
 
-	// Run the server
-	err := asyncServer.Run(ctx)
+	shardManagerCtx, cancelShardManager := context.WithCancel(ctx)
+	defer cancelShardManager()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		shardManager.Run(shardManagerCtx)
+	}()
+
+	serverErrCh := make(chan error, 2)
+	var serverWg sync.WaitGroup
+	serverWg.Add(1)
+	go func() {
+		defer serverWg.Done()
+		// Run the server
+		err := asyncServer.Run(ctx)
+
+		// Handling different server errors
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				log.Info("Server was canceled")
+			} else if errors.Is(err, server.ErrAborted) {
+				log.Info("Server received abort command")
+			} else {
+				log.Error("Server error", "error", err)
+			}
+			serverErrCh <- err
+		} else {
+			log.Info("Server stopped without error")
+		}
+	}()
 
-	// Handling different server errors
-	if err != nil {
-		if errors.Is(err, context.Canceled) {
-			log.Debug("Server was canceled")
-		} else if errors.Is(err, server.ErrAborted) {
-			log.Debug("Server received abort command")
+	serverWg.Add(1)
+	go func() {
+		defer serverWg.Done()
+		// Run the HTTP server
+		err := httpServer.Run(ctx)
```

Comment:
Should we create a cancellable context from the ctx for the `httpServer` and `asyncServer` to maintain consistency with the shard manager? Or change the shard manager to use the common ctx.

(Actually I would like to know what the preferred practice is)

Reply:
So, we create a child context when we need to maintain a separate lifecycle, i.e we can cancel the child context without impacting the parent.
Any issues with the Async or HTTP server would mean we want to shut things down and cancel all child contexts. I don't think that we would like to keep Shard Manager at the common context level.

We can, however, create separate contexts for each of them, but we need to remember that each cancellable context spawns a goroutine to handle cancellation. We're just talking about two more, so it should be fine.